### PR TITLE
Implementation of load average calculation and corresponding /proc entry

### PR DESCRIPTION
**Pull request by Champii**
#### (First contribution)

This PR introduces the load average calculations based on how linux does it.

It adds:
  - A `trait SchedulerStats` with a method `nr_queued_and_running()`  that return the number of queued tasks and the number of actually running tasks on all CPUs. 
  - An implementation of this trait for the `PreemptScheduler`
  - A way to get this value from a `SCHEDULER_STATS` singleton via a similarly named public function.
  - A timer callback to periodically trigger the loadavg update
  - A new function to get the last allocated tid in `posix_thread::last_tid()`
 
This is then used to populate a new `/proc` entry named `loadavg`

### CODE REVIEW

**Comment by tatetian** [comment]
@Champii Welcome to the Asterinas community! Good job on implementing the load average feature. This statistic is very useful to the user space. I am curious why you chose this feature as your first target. Mind to share your thoughts?

Back to the technical aspect of this PR. The current form of this PR is not ideal. Let me explain why.

The thing is that in Asterinas, we take memory safety as a top priority and try very hard to minimize the TCB in terms of memory safety. We have a very clear TCB, which is the `ostd` crate. Every line of code in `ostd` is viewed as part of the TCB - even if it is safe Rust code! That is one of the main reason that we enable an `ostd`-based kernel to _inject_ a custom task scheduler back into `ostd`. A task scheduler could be arbitrarily complex and we do not want that complexity wired into `ostd`. Even the built-in, minimalism FIFO scheduler in `ostd` will be soon moved out of `ostd`.

Another main reason of supporting task scheduler injection is _flexibility_. The `ostd` crate as well as the OSDK tool are designed to be _OS-neutral_. We want to enable other developers to leverage OSDK and OSTD to build new Rust-based OS that have different tradeoffs than Asterinas. We do not want OSTD to contain any features specific to Linux. So neither Linux's Completely Fair Scheduler (CFS) nor the load average feature should be baked into OSTD.

Now that the design considerations are explained, I think it should now be clear to you what this PR should look like. The load average feature is specific to Linux. So almost all changes to support this feature should be made to `kernel/src`, not `ostd`. The `Scheduler` trait should NOT be touched. Nor should a `loadavg` module added to `ostd`. All new counters and statistics needed for calculating load averages should be obtained from or maintained by the concrete scheduler implementation at `kernel/src/sched/priority_scheduler.rs`.

**Comment by Champii** [comment]
> @Champii Welcome to the Asterinas community! Good job on implementing the load average feature. This statistic is very useful to the user space. I am curious why you chose this feature as your first target. Mind to share your thoughts?

Hello @tatetian, thank you for this warm welcome !  
I've first looked at the missing syscalls from the list you have on the website, and tried to find one that would be a good first addition for a newcomer. I've chosen `sysinfo` because it looked simple enough, and to implement it I would need to implement the load average as well as the swap management. Out of those two steps I chose to start with the load average.

> Back to the technical aspect of this PR. The current form of this PR is not ideal. Let me explain why.

This makes total sense, thank you for this explanation :)  
I'll rework this PR this week to have everything implemented on the `kernel/src` part

Amazing work btw, I love what you guys are doing

File: kernel/src/sched/priority_scheduler.rs
Comment (junyang-zh) [inline kernel/src/sched/priority_scheduler.rs:28]: We have used the word "inject" to refer to something not implemented by OSTD but set as a system-wide singleton to make part of the OSTD's job work. In this case I think it is better phrased by "`set_stats_from_scheduler`".

File: kernel/src/sched/stats/loadavg.rs
Comment (junyang-zh) [inline kernel/src/sched/stats/loadavg.rs:36]: The general practice is not to use `lazy_static`, as it's both discouraged by the Rust community and have problems in our system due to the difference in the implementation of the synchronization.

In this case, `LoadAverage` can clearly have a `const new()` function that can serve static initialization.

File: kernel/src/sched/priority_scheduler.rs
Comment (junyang-zh) [inline kernel/src/sched/priority_scheduler.rs:160]: When the load average is queried through the proc file system, the two functions will be called one by one. And they will respectively lock the queues twice and unlock them twice. So there may be discrepancies if the current task is preempted in between the two calls. I think a good design of the trait won't introduce such problem. A suggestion is to return these two stats in one function.

It is a minor suggestion and feel free to argue with me.

**Review by junyang-zh — COMMENTED** [review]
Looks promising! And thanks for your willingness to help in the Asterinas community. There are some suggestions.

File: kernel/src/sched/stats/loadavg.rs
Comment (Champii) [reply kernel/src/sched/stats/loadavg.rs:36]: It can indeed, good catch

File: kernel/src/sched/priority_scheduler.rs
Comment (Champii) [reply kernel/src/sched/priority_scheduler.rs:160]: I believe you have a good point here. I'm actually thinking about what other things would need to get those statistics and if we would benefit from keeping them separated (maybe to save some precious cycles in a performance sensitive context ?)

It is to be noted that the linux kernel itself doesn't seem to care so much about the load average accuracy, and if this data is only used to actually print some stats to the user we could as well tolerate some (rare) inaccuracy.

You definitely have a better big picture of the whole system so I will rely on your expertise :)

**Comment by junyang-zh** [comment]
The lint CI failed because of formatting issues. You can type `make format` to automatically format Rust files for you. :>

**Comment by Champii** [comment]
My bad, I should setup a git hook for that

File: kernel/src/sched/priority_scheduler.rs
Comment (Champii) [reply kernel/src/sched/priority_scheduler.rs]: I merged those methods together. It does make more sense

File: kernel/src/sched/stats/loadavg.rs
Comment (tatetian) [inline kernel/src/sched/stats/loadavg.rs:57]: In kernel programming, we cannot use floating point numbers. For one thing, when doing user-kernel switching (e.g., system calls), the floating pointer portion of the CPU state is not saved for efficiency. So the kernel should NOT touch the floating pointer state by doing floating pointer arithmetic.

Furthermore, after importing the `fixed` crate, the fixed point types provided by it support `Debug` and `Display` traits (I guess). So you do not have to convert fixed-point numbers to floating-point ones in order to display them as a string.

File: kernel/src/sched/stats/loadavg.rs
Comment (tatetian) [inline kernel/src/sched/stats/loadavg.rs:73]: Most calls to `update_loadavg` will not do any real work.

```rust
    timer::register_callback(|| {
        loadavg::update_loadavg(nr_queued_and_running());
    });
```

But every call to `update_loadavg` obtains data from the scheduler, which involves locking and iterating over every runqueues. So to improve efficiency, I suggest refactoring the interface of `update_loadavg` to take a function pointer so that `nr_queued_and_running` can be called lazily.

File: kernel/src/sched/stats/loadavg.rs
Comment (tatetian) [inline kernel/src/sched/stats/loadavg.rs:72]: To make this fast path even faster, extract the filed `last_update` out of `LOAD_AVG` as an atomic variable so that you don't have to lock `LOAD_AVG` in order to read an integer.

**Review by tatetian — COMMENTED** [review]
Good job on keeping the modifications to the `kernel` directory!

File: kernel/src/sched/stats/loadavg.rs
Comment (Champii) [reply kernel/src/sched/stats/loadavg.rs:73]: This is smart, I was worried about the efficiency of my implementation. Thank you :)

File: kernel/src/syscall/fcntl.rs
Comment (Champii) [inline kernel/src/syscall/fcntl.rs:49]: I had to specify the type here

File: kernel/src/fs/utils/page_cache.rs
Comment (Champii) [inline kernel/src/fs/utils/page_cache.rs:377]: And here I had to dereference the usize

**Review by Champii — COMMENTED** [review]
I successfully used `fixed` instead of the custom solution I implemented, but for unknown reasons it made the compilation failed in two completely different modules. 

I tried to reproduce a minimal case where this could happen but without success for now. We might want to look into that because I'm not super comfortable with just changing some unrelated code without understanding why this is happening.  

If you have any ideas I'm listening :)

**Comment by Champii** [comment]
Small update on my progress on this issue, I've made a minimal program that reproduce this behavior

With `fixed = "1.28.0"`

```rust
fn main() {}

type FixedPoint = fixed::types::U21F11;

fn test(arg: u64) {
    arg > u8::MAX.into();
}
```

```
error[E0283]: type annotations needed
 --> src/main.rs:8:19
  |
8 |     arg > u8::MAX.into();
  |         -         ^^^^
  |         |
  |         type must be known at this point
  |
  = note: cannot satisfy `u64: PartialOrd<_>`
  = help: the following types implement trait `PartialOrd<Rhs>`:
            <u64 as PartialOrd<FixedI128<Frac>>>
            <u64 as PartialOrd<FixedI16<Frac>>>
            <u64 as PartialOrd<FixedI32<Frac>>>
            <u64 as PartialOrd<FixedI64<Frac>>>
            <u64 as PartialOrd<FixedI8<Frac>>>
            <u64 as PartialOrd<FixedU128<Frac>>>
            <u64 as PartialOrd<FixedU16<Frac>>>
            <u64 as PartialOrd<FixedU32<Frac>>>
          and 3 others
help: try using a fully qualified path to specify the expected types
  |
8 |     arg > <u8 as Into<T>>::into(u8::MAX);
  |           ++++++++++++++++++++++       ~

For more information about this error, try `rustc --explain E0283`.
error: could not compile `lol` (bin "lol") due to 1 previous error
```

If I comment the `type FixedPoint = fixed::types::U21F11;` then the program compiles fine. It seems that by just pulling the `fixed` type into the scope it messes up the compiler typechecking.

I'll open an issue on `fixed` to understand what is going on and hopefully propose a fix

**Comment by tatetian** [comment]
This is indeed surprising, but explainable. The `fixed` crate introduces multiple new types (`PartialOrd`) that may be compared with `u64`. So the compiler cannot infer a single type to satisfy the `PartialOrd` trait.

See [the minimal example on Rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=392529c03efd3bde57ddc980b243d604).

**Comment by Champii** [comment]
> This is indeed surprising, but explainable. The `fixed` crate introduces multiple new types (`PartialOrd`) that may be compared with `u64`. So the compiler cannot infer a single type to satisfy the `PartialOrd` trait.
> 
> See [the minimal example on Rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=392529c03efd3bde57ddc980b243d604).

Yes that makes total sense, thank you :)

Then we have two options:
 - Either to embrace the addition of those impl traits over primitive types (this is the current state of this PR) 
 - Or we decide to go with another lib that does the same thing

I'll lean on your decision about this

**Comment by tatetian** [comment]
> Then we have two options:
> 
> * Either to embrace the addition of those impl traits over primitive types (this is the current state of this PR)
> * Or we decide to go with another lib that does the same thing

I think the first option is acceptable.

**Comment by Champii** [comment]
Awesome, I just rebased from `main`, and I believe this PR is good to go if you don't see anything else

File: kernel/src/sched/stats/loadavg.rs
Comment (tatetian) [inline kernel/src/sched/stats/loadavg.rs:70]: Rename `actual_load` to `old_load`, which is aligned with the new `new_load`.

File: kernel/Cargo.toml
Comment (tatetian) [inline kernel/Cargo.toml:62]: Add a TODO here.

```rust
# Fixed point numbers
# TODO: fork this crate to rewrite all the (unnecessary) unsafe usage
fixed = "1.28.0"
```

File: kernel/src/sched/stats/loadavg.rs
Comment (tatetian) [inline kernel/src/sched/stats/loadavg.rs:69]: Please remove this comment. The code below is to calculate the sliding window averages; it is not unique to Linux. And the code is definitely different from that of Linux. So we are not copying or taking code from Linux here.

As the GPL license of Linux is incompatible with the MPL license of Asterinas, we are not allowed to _take_ or _copy_ code from Linux. Instead we should _learn_ or _steal_ ideas from Linux.

**Review by tatetian — COMMENTED** [review]
The PR looks mostly good. Just a few minor issues. After fixing them, we will be good to merge it.

**Comment by tatetian** [comment]
@boterinas rerun

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**
